### PR TITLE
Implemented bulk read/write with BatchGet/Write APIs

### LINKED ISSUE #114: Implement DynamoDB bulk operations
Implement the following using Batch*Item APIs:

- [x] bulkWrite with BatchWriteItem
- [x] bulkRead with BatchGetItem

### PR DESCRIPTION
**Pull request by amcp**
Fixes #114

### CODE REVIEW

**Comment by amcp** [comment]
@vinaykumarchella I no longer special case throttling and ISE here.

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 1]
```diff
@@ -188,13 +202,97 @@ public String writeSingle(String key) throws Exception {
...
+		Preconditions.checkArgument(new HashSet<>(keys).size() == keys.size());
+		final KeysAndAttributes keysAndAttributes = generateReadRequests(keys);
+		try {
+			readUntilDone(keysAndAttributes);
+			return keysAndAttributes.getKeys().stream()
+					.map(Map::toString)
+					.collect(Collectors.toList());
+		} catch (AmazonServiceException ase) {
+			amazonServiceException(ase);
+			return new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: fixed

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 2]
```diff
@@ -188,13 +202,97 @@ public String writeSingle(String key) throws Exception {
...
+		final List<WriteRequest> writeRequests = generateWriteRequests(keys);
+		try {
+			writeUntilDone(writeRequests);
+			return writeRequests.stream()
+					.map(WriteRequest::getPutRequest)
+					.map(PutRequest::toString)
+					.collect(Collectors.toList());
+		} catch (AmazonServiceException ase) {
+			amazonServiceException(ase);
+			return new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: fixed

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 3]
```diff
@@ -188,12 +202,96 @@ public String writeSingle(String key) throws Exception {
...
+        BatchWriteItemResult result;
+        do {
+            result = runBatchWriteRequest(remainingRequests);
+            remainingRequests = result.getUnprocessedItems().get(table.getTableName());
+        } while (remainingRequests!= null && remainingRequests.isEmpty());
+    }
+
+    private BatchWriteItemResult runBatchWriteRequest(List<WriteRequest> writeRequests) {
+        //estimate size of requests
+        int totalSize = writeRequests.stream()
```
Comment (vinaykumarchella) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: Are you using `totalSize` anywhere?

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 4]
```diff
@@ -79,19 +94,18 @@
...
-            logger.error("Discovery Environment Variable: " + discoveryEnv);
+        String discoveryEnv = System.getenv(NdBenchConstants.DISCOVERY_ENV);
+        logger.error("Discovery Environment Variable: " + discoveryEnv);
         if (discoveryEnv == null || discoveryEnv.equals(NdBenchConstants.DISCOVERY_ENV_AWS)) {
             awsCredentialsProvider = credential;
         } else {
             awsCredentialsProvider = new ProfileCredentialsProvider();
             try {
-
-            awsCredentialsProvider.getCredentials();
```
Comment (vinaykumarchella) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: Should the credentials loading part be part of `init`?

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java (hunk 3 above)
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: For variable data sizes in config, it is a good practice to self-throttle on the client side based on the expected WCU consumption of each item. I haven't implemented this yet, but the size calculation is a precursor: ceiling(bytes / 1024) = WCU. I can pull this out and save for a later date.

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java (hunk 4 above)
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: Can we discuss the location of credentials in an issue to resolve in a different PR? This is a non-essential spacing change.

Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: #144 opened

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java (hunk 3 above)
Comment (amcp) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: fixed

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java (hunk 4 above)
Comment (vinaykumarchella) [reply ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: 👍

**Comment by vinaykumarchella** [comment]
Can you rebase this change.

**Comment by amcp** [comment]
Rebased.

**Pull request merged**